File: snippets/design-patterns/factory/factory_pattern.js
/**
 * Factory Pattern Implementation in JavaScript
 *
 * The Factory Pattern is a creational design pattern that provides an interface for creating
 * objects in a superclass, but allows subclasses to alter the type of objects that will be created.
 *
 * This example demonstrates a Vehicle Factory that can create different types of vehicles
 * (Car, Motorcycle, Truck) based on the client's requirements.
 */

// Abstract Product - Vehicle
class Vehicle {
  constructor(make, model, year) {
    this.make = make;
    this.model = model;
    this.year = year;
  }

  getInfo() {
    return `${this.year} ${this.make} ${this.model}`;
  }

  start() {
    return `${this.getInfo()} is starting...`;
  }

  stop() {
    return `${this.getInfo()} is stopping...`;
  }
}

// Concrete Products
class Car extends Vehicle {
  constructor(make, model, year, doors = 4) {
    super(make, model, year);
    this.doors = doors;
    this.type = 'Car';
  }

  getInfo() {
    return `${super.getInfo()} (${this.doors}-door car)`;
  }

  drive() {
    return `${this.getInfo()} is driving on the road.`;
  }
}

class Motorcycle extends Vehicle {
  constructor(make, model, year, engineSize) {
    super(make, model, year);
    this.engineSize = engineSize;
    this.type = 'Motorcycle';
  }

  getInfo() {
    return `${super.getInfo()} (${this.engineSize}cc motorcycle)`;
  }

  ride() {
    return `${this.getInfo()} is riding at high speed.`;
  }
}

class Truck extends Vehicle {
  constructor(make, model, year, capacity) {
    super(make, model, year);
    this.capacity = capacity;
    this.type = 'Truck';
  }

  getInfo() {
    return `${super.getInfo()} (${this.capacity} ton truck)`;
  }

  haul() {
    return `${this.getInfo()} is hauling cargo.`;
  }
}

// Simple Factory
class VehicleFactory {
  static createVehicle(type, make, model, year, options = {}) {
    switch (type.toLowerCase()) {
      case 'car':
        return new Car(make, model, year, options.doors);
      case 'motorcycle':
        return new Motorcycle(make, model, year, options.engineSize);
      case 'truck':
        return new Truck(make, model, year, options.capacity);
      default:
        throw new Error(`Vehicle type ${type} is not supported.`);
    }
  }
}

// Factory Method Pattern Implementation
class VehicleFactoryMethod {
  createVehicle(make, model, year, options = {}) {
    // This method will be implemented by concrete factories
    throw new Error('createVehicle method must be implemented by subclasses');
  }

  registerVehicle(make, model, year, options = {}) {
    // Common operations for all vehicles
    const vehicle = this.createVehicle(make, model, year, options);
    console.log(`Registering ${vehicle.getInfo()}`);
    console.log(`Assigning license plate to ${vehicle.type}`);
    return vehicle;
  }
}

// Concrete Factories
class CarFactory extends VehicleFactoryMethod {
  createVehicle(make, model, year, options = {}) {
    return new Car(make, model, year, options.doors || 4);
  }
}

class MotorcycleFactory extends VehicleFactoryMethod {
  createVehicle(make, model, year, options = {}) {
    return new Motorcycle(make, model, year, options.engineSize || 250);
  }
}

class TruckFactory extends VehicleFactoryMethod {
  createVehicle(make, model, year, options = {}) {
    return new Truck(make, model, year, options.capacity || 5);
  }
}

// Abstract Factory Pattern Implementation
class VehiclePartsFactory {
  createEngine() {
    throw new Error('Method not implemented');
  }

  createTransmission() {
    throw new Error('Method not implemented');
  }

  createChassis() {
    throw new Error('Method not implemented');
  }
}

// Parts
class Engine {
  constructor(type, horsepower) {
    this.type = type;
    this.horsepower = horsepower;
  }

  getSpecs() {
    return `${this.type} engine with ${this.horsepower}hp`;
  }
}

class Transmission {
  constructor(type, gears) {
    this.type = type;
    this.gears = gears;
  }

  getSpecs() {
    return `${this.type} transmission with ${this.gears} gears`;
  }
}

class Chassis {
  constructor(material, weight) {
    this.material = material;
    this.weight = weight;
  }

  getSpecs() {
    return `${this.material} chassis weighing ${this.weight}kg`;
  }
}

// Concrete Abstract Factories
class SportVehiclePartsFactory extends VehiclePartsFactory {
  createEngine() {
    return new Engine('V8', 450);
  }

  createTransmission() {
    return new Transmission('Manual', 6);
  }

  createChassis() {
    return new Chassis('Carbon Fiber', 120);
  }
}

class EconomyVehiclePartsFactory extends VehiclePartsFactory {
  createEngine() {
    return new Engine('Inline-4', 180);
  }

  createTransmission() {
    return new Transmission('Automatic', 5);
  }

  createChassis() {
    return new Chassis('Steel', 300);
  }
}

class HeavyDutyVehiclePartsFactory extends VehiclePartsFactory {
  createEngine() {
    return new Engine('Diesel V6', 350);
  }

  createTransmission() {
    return new Transmission('Manual', 8);
  }

  createChassis() {
    return new Chassis('Reinforced Steel', 800);
  }
}

// Vehicle Assembler - Uses the Abstract Factory
class VehicleAssembler {
  constructor(partsFactory) {
    this.partsFactory = partsFactory;
  }

  assembleVehicle() {
    const engine = this.partsFactory.createEngine();
    const transmission = this.partsFactory.createTransmission();
    const chassis = this.partsFactory.createChassis();

    console.log('Assembling vehicle with:');
    console.log(`- ${engine.getSpecs()}`);
    console.log(`- ${transmission.getSpecs()}`);
    console.log(`- ${chassis.getSpecs()}`);

    return {
      engine,
      transmission,
      chassis
    };
  }
}

// Client code demonstration
function clientCode() {
  console.log('===== Simple Factory Pattern =====');

  const car = VehicleFactory.createVehicle('car', 'Toyota', 'Camry', 2023, { doors: 4 });
  const motorcycle = VehicleFactory.createVehicle('motorcycle', 'Honda', 'CBR', 2023, { engineSize: 600 });
  const truck = VehicleFactory.createVehicle('truck', 'Ford', 'F-150', 2023, { capacity: 3 });

  console.log(car.getInfo());
  console.log(car.drive());

  console.log(motorcycle.getInfo());
  console.log(motorcycle.ride());

  console.log(truck.getInfo());
  console.log(truck.haul());

  console.log('\n===== Factory Method Pattern =====');

  const carFactory = new CarFactory();
  const motorcycleFactory = new MotorcycleFactory();
  const truckFactory = new TruckFactory();

  const newCar = carFactory.registerVehicle('BMW', '3 Series', 2023, { doors: 2 });
  const newMotorcycle = motorcycleFactory.registerVehicle('Ducati', 'Monster', 2023, { engineSize: 821 });
  const newTruck = truckFactory.registerVehicle('Volvo', 'VNL', 2023, { capacity: 20 });

  console.log(newCar.drive());
  console.log(newMotorcycle.ride());
  console.log(newTruck.haul());

  console.log('\n===== Abstract Factory Pattern =====');

  const sportPartsFactory = new SportVehiclePartsFactory();
  const economyPartsFactory = new EconomyVehiclePartsFactory();
  const heavyDutyPartsFactory = new HeavyDutyVehiclePartsFactory();

  console.log('Building a sports car:');
  const sportCarAssembler = new VehicleAssembler(sportPartsFactory);
  const sportCarParts = sportCarAssembler.assembleVehicle();

  console.log('\nBuilding an economy car:');
  const economyCarAssembler = new VehicleAssembler(economyPartsFactory);
  const economyCarParts = economyCarAssembler.assembleVehicle();

  console.log('\nBuilding a heavy duty truck:');
  const heavyDutyTruckAssembler = new VehicleAssembler(heavyDutyPartsFactory);
  const heavyDutyTruckParts = heavyDutyTruckAssembler.assembleVehicle();
}

// Execute the client code
// Run the example
clientCode();

// Export classes for use in other modules
module.exports = {
  // Simple Factory
  VehicleFactory,

  // Factory Method
  VehicleFactoryMethod,
  CarFactory,
  MotorcycleFactory,
  TruckFactory,

  // Abstract Factory
  VehiclePartsFactory,
  SportVehiclePartsFactory,
  EconomyVehiclePartsFactory,
  HeavyDutyVehiclePartsFactory,
  VehicleAssembler,

  // Products
  Vehicle,
  Car,
  Motorcycle,
  Truck,
  Engine,
  Transmission,
  Chassis
};

File: snippets/design-patterns/factory/FactoryPattern.cs
/**
 * Factory Pattern Implementation in C#
 *
 * The Factory Pattern is a creational design pattern that provides an interface for creating
 * objects in a superclass, but allows subclasses to alter the type of objects that will be created.
 *
 * This example demonstrates a Vehicle Factory that can create different types of vehicles
 * (Car, Motorcycle, Truck) based on the client's requirements.
 */

using System;
using System.Collections.Generic;

namespace DesignPatterns.Factory
{
    // Abstract Product - Vehicle
    public abstract class Vehicle
    {
        protected string Make { get; }
        protected string Model { get; }
        protected int Year { get; }

        protected Vehicle(string make, string model, int year)
        {
            Make = make;
            Model = model;
            Year = year;
        }

        public virtual string GetInfo()
        {
            return $"{Year} {Make} {Model}";
        }

        public virtual string Start()
        {
            return $"{GetInfo()} is starting...";
        }

        public virtual string Stop()
        {
            return $"{GetInfo()} is stopping...";
        }
    }

    // Concrete Products
    public class Car : Vehicle
    {
        public int Doors { get; }

        public Car(string make, string model, int year, int doors = 4)
            : base(make, model, year)
        {
            Doors = doors;
        }

        public override string GetInfo()
        {
            return $"{base.GetInfo()} ({Doors}-door car)";
        }

        public string Drive()
        {
            return $"{GetInfo()} is driving on the road.";
        }
    }

    public class Motorcycle : Vehicle
    {
        public int EngineSize { get; }

        public Motorcycle(string make, string model, int year, int engineSize)
            : base(make, model, year)
        {
            EngineSize = engineSize;
        }

        public override string GetInfo()
        {
            return $"{base.GetInfo()} ({EngineSize}cc motorcycle)";
        }

        public string Ride()
        {
            return $"{GetInfo()} is riding at high speed.";
        }
    }

    public class Truck : Vehicle
    {
        public double Capacity { get; }

        public Truck(string make, string model, int year, double capacity)
            : base(make, model, year)
        {
            Capacity = capacity;
        }

        public override string GetInfo()
        {
            return $"{base.GetInfo()} ({Capacity} ton truck)";
        }

        public string Haul()
        {
            return $"{GetInfo()} is hauling cargo.";
        }
    }

    // Simple Factory
    public class VehicleFactory
    {
        public enum VehicleType
        {
            Car,
            Motorcycle,
            Truck
        }

        public static Vehicle CreateVehicle(VehicleType type, string make, string model, int year, object options = null)
        {
            switch (type)
            {
                case VehicleType.Car:
                    int doors = 4;
                    if (options is int doorsOption)
                    {
                        doors = doorsOption;
                    }
                    return new Car(make, model, year, doors);

                case VehicleType.Motorcycle:
                    int engineSize = 250;
                    if (options is int engineSizeOption)
                    {
                        engineSize = engineSizeOption;
                    }
                    return new Motorcycle(make, model, year, engineSize);

                case VehicleType.Truck:
                    double capacity = 5.0;
                    if (options is double capacityOption)
                    {
                        capacity = capacityOption;
                    }
                    return new Truck(make, model, year, capacity);

                default:
                    throw new ArgumentException($"Vehicle type {type} is not supported.");
            }
        }
    }

    // Factory Method Pattern Implementation
    public abstract class VehicleFactoryMethod
    {
        public abstract Vehicle CreateVehicle(string make, string model, int year, object options = null);

        public Vehicle RegisterVehicle(string make, string model, int year, object options = null)
        {
            // Common operations for all vehicles
            Vehicle vehicle = CreateVehicle(make, model, year, options);
            Console.WriteLine($"Registering {vehicle.GetInfo()}");
            Console.WriteLine("Assigning license plate");
            return vehicle;
        }
    }

    // Concrete Factories
    public class CarFactory : VehicleFactoryMethod
    {
        public override Vehicle CreateVehicle(string make, string model, int year, object options = null)
        {
            int doors = 4;
            if (options is int doorsOption)
            {
                doors = doorsOption;
            }
            return new Car(make, model, year, doors);
        }
    }

    public class MotorcycleFactory : VehicleFactoryMethod
    {
        public override Vehicle CreateVehicle(string make, string model, int year, object options = null)
        {
            int engineSize = 250;
            if (options is int engineSizeOption)
            {
                engineSize = engineSizeOption;
            }
            return new Motorcycle(make, model, year, engineSize);
        }
    }

    public class TruckFactory : VehicleFactoryMethod
    {
        public override Vehicle CreateVehicle(string make, string model, int year, object options = null)
        {
            double capacity = 5.0;
            if (options is double capacityOption)
            {
                capacity = capacityOption;
            }
            return new Truck(make, model, year, capacity);
        }
    }

    // Abstract Factory Pattern Implementation
    // Parts
    public class Engine
    {
        public string Type { get; }
        public int Horsepower { get; }

        public Engine(string type, int horsepower)
        {
            Type = type;
            Horsepower = horsepower;
        }

        public string GetSpecs()
        {
            return $"{Type} engine with {Horsepower}hp";
        }
    }

    public class Transmission
    {
        public string Type { get; }
        public int Gears { get; }

        public Transmission(string type, int gears)
        {
            Type = type;
            Gears = gears;
        }

        public string GetSpecs()
        {
            return $"{Type} transmission with {Gears} gears";
        }
    }

    public class Chassis
    {
        public string Material { get; }
        public double Weight { get; }

        public Chassis(string material, double weight)
        {
            Material = material;
            Weight = weight;
        }

        public string GetSpecs()
        {
            return $"{Material} chassis weighing {Weight}kg";
        }
    }

    // Abstract Factory
    public interface IVehiclePartsFactory
    {
        Engine CreateEngine();
        Transmission CreateTransmission();
        Chassis CreateChassis();
    }

    // Concrete Abstract Factories
    public class SportVehiclePartsFactory : IVehiclePartsFactory
    {
        public Engine CreateEngine()
        {
            return new Engine("V8", 450);
        }

        public Transmission CreateTransmission()
        {
            return new Transmission("Manual", 6);
        }

        public Chassis CreateChassis()
        {
            return new Chassis("Carbon Fiber", 120);
        }
    }

    public class EconomyVehiclePartsFactory : IVehiclePartsFactory
    {
        public Engine CreateEngine()
        {
            return new Engine("Inline-4", 180);
        }

        public Transmission CreateTransmission()
        {
            return new Transmission("Automatic", 5);
        }

        public Chassis CreateChassis()
        {
            return new Chassis("Steel", 300);
        }
    }

    public class HeavyDutyVehiclePartsFactory : IVehiclePartsFactory
    {
        public Engine CreateEngine()
        {
            return new Engine("Diesel V6", 350);
        }

        public Transmission CreateTransmission()
        {
            return new Transmission("Manual", 8);
        }

        public Chassis CreateChassis()
        {
            return new Chassis("Reinforced Steel", 800);
        }
    }

    // Vehicle Assembler - Uses the Abstract Factory
    public class VehicleAssembler
    {
        private readonly IVehiclePartsFactory _partsFactory;

        public VehicleAssembler(IVehiclePartsFactory partsFactory)
        {
            _partsFactory = partsFactory;
        }

        public void AssembleVehicle()
        {
            Engine engine = _partsFactory.CreateEngine();
            Transmission transmission = _partsFactory.CreateTransmission();
            Chassis chassis = _partsFactory.CreateChassis();

            Console.WriteLine("Assembling vehicle with:");
            Console.WriteLine($"- {engine.GetSpecs()}");
            Console.WriteLine($"- {transmission.GetSpecs()}");
            Console.WriteLine($"- {chassis.GetSpecs()}");
        }
    }

    // Client code demonstration
    public class Client
    {
        public static void DemonstrateFactoryPattern()
        {
            Console.WriteLine("===== Simple Factory Pattern =====");

            Vehicle car = VehicleFactory.CreateVehicle(VehicleFactory.VehicleType.Car, "Toyota", "Camry", 2023, 4);
            Vehicle motorcycle = VehicleFactory.CreateVehicle(VehicleFactory.VehicleType.Motorcycle, "Honda", "CBR", 2023, 600);
            Vehicle truck = VehicleFactory.CreateVehicle(VehicleFactory.VehicleType.Truck, "Ford", "F-150", 2023, 3.0);

            Console.WriteLine(car.GetInfo());
            Console.WriteLine(((Car)car).Drive());

            Console.WriteLine(motorcycle.GetInfo());
            Console.WriteLine(((Motorcycle)motorcycle).Ride());

            Console.WriteLine(truck.GetInfo());
            Console.WriteLine(((Truck)truck).Haul());

            Console.WriteLine("\n===== Factory Method Pattern =====");

            CarFactory carFactory = new CarFactory();
            MotorcycleFactory motorcycleFactory = new MotorcycleFactory();
            TruckFactory truckFactory = new TruckFactory();

            Vehicle newCar = carFactory.RegisterVehicle("BMW", "3 Series", 2023, 2);
            Vehicle newMotorcycle = motorcycleFactory.RegisterVehicle("Ducati", "Monster", 2023, 821);
            Vehicle newTruck = truckFactory.RegisterVehicle("Volvo", "VNL", 2023, 20.0);

            Console.WriteLine(((Car)newCar).Drive());
            Console.WriteLine(((Motorcycle)newMotorcycle).Ride());
            Console.WriteLine(((Truck)newTruck).Haul());

            Console.WriteLine("\n===== Abstract Factory Pattern =====");

            Console.WriteLine("Building a sports car:");
            VehicleAssembler sportCarAssembler = new VehicleAssembler(new SportVehiclePartsFactory());
            sportCarAssembler.AssembleVehicle();

            Console.WriteLine("\nBuilding an economy car:");
            VehicleAssembler economyCarAssembler = new VehicleAssembler(new EconomyVehiclePartsFactory());
            economyCarAssembler.AssembleVehicle();

            Console.WriteLine("\nBuilding a heavy duty truck:");
            VehicleAssembler heavyDutyTruckAssembler = new VehicleAssembler(new HeavyDutyVehiclePartsFactory());
            heavyDutyTruckAssembler.AssembleVehicle();
        }
    }

    // Main program, run the example
    class Program
    {
        static void Main(string[] args)
        {
            Client.DemonstrateFactoryPattern();
            Console.ReadLine();
        }
    }
}
